File: longest_palindromic_subsequence.cpp
//{ Driver Code Starts
// Initial template for C++

#include <bits/stdc++.h>
using namespace std;

// } Driver Code Ends
// User function Template for C++

class Solution {
  int dp[1001][1001];
  string s;
  int solve(int i, int j) {
    if (i > j) return 0;
    if (i == j) return 1;
    if (dp[i][j] != -1) return dp[i][j];
    if (s[i] == s[j]) return dp[i][j] = solve(i + 1, j - 1) + 2;
    return dp[i][j] = max(solve(i + 1, j), solve(i, j - 1));
  }

 public:
  int longestPalinSubseq(string A) {
    // code here
    memset(dp, -1, sizeof(dp));
    s = A;
    return solve(0, A.size() - 1);
  }
};

//{ Driver Code Starts.

int32_t main() {
  int t;
  cin >> t;
  while (t--) {
    string s;
    cin >> s;
    Solution ob;
    cout << ob.longestPalinSubseq(s) << endl;
  }
}
// Contributed By: Pranay Bansal

// } Driver Code Ends

File: dp/longest_palindromic_subsequence.py
# User function Template for python3
import sys

sys.setrecursionlimit(2000)


class Solution:
    def longestPalinSubseq(self, S):
        # code here
        n = len(S)
        dp = [[-1 for _ in range(n)] for _ in range(n)]

        def f(i, j):
            if i > j:
                return 0
            if i == j:
                return 1
            if dp[i][j] != -1:
                return dp[i][j]

            if S[i] == S[j]:
                dp[i][j] = 2 + f(i + 1, j - 1)
            else:
                dp[i][j] = max(f(i + 1, j), f(i, j - 1))
            return dp[i][j]

        return f(0, n - 1)


# {
# Driver Code Starts
# Initial Template for Python 3

if __name__ == "__main__":
    t = int(input())
    for _ in range(t):
        s = input()
        ob = Solution()
        ans = ob.longestPalinSubseq(s)
        print(ans)
# } Driver Code Ends
